Fix for channel names containing a forward slash

StollD opened the pull request:
Its me again

@eggrobin discovered that the log viewer breaks on channel names containing a forward slash, because they don't get escaped. This pull request handles (or at least tries to) this.

Additionally, we (or rather eggrobin) detected another bug. When a channel name starts with a . after the #, the logger won't be able to see a difference between #.channelname and ##channelname. Of course this could be changed, by simply leaving the second # there, and not changing it to a ., but this would break the links for this channel on your logging instance: https://irclog.whitequark.org/.openfpga/2017-03-05
As they might use the logs through the link, breaking it might be a bit unfortunate.

Do you have any ideas on this?

whitequark commented on lib/irclogger/viewer_helpers.rb:
```diff
@@ -5,11 +5,11 @@ module ViewerHelpers
 
     def channel_escape(channel)
-      channel[1..-1].gsub(/^#+/) { |m| '.' * m.length }
+      CGI.escape(channel[1..-1]).gsub(/^#+/) { |m| '.' * m.length }
```
So... what does the slash become? Some %-encoded atrocity? How about replacing it with a `⁄`?

whitequark commented:
> Of course this could be changed, by simply leaving the second # there

It cannot--a # signifies an anchor in an URL. And no, breaking old links isn't an option.

eggrobin commented:
@StollD is asleep (and so should I be), but we discussed the options before he made this pull request: by "leaving the second # there" he means having it escaped by whichever escaping scheme is chosen. Of course that does have the problem of old links, see the following paragraph for my thoughts on a workaround.
Percent-encoding has the advantage of being somewhat standard, but is admittedly rather awful; I like the idea of Unicode tricks; in order to allow this to work in general (I am in a surprising number of channels with strange things in their names), the fullwidth forms (U+FF01–FF5E) may be an option.

In order to preserve old links, prefix strings of `.` (U+2E FULL STOP) in links should be mapped to chains of `#` (U+23 NUMBER SIGN) in channel names; that means that something else must be used in the the links to logs of a channel whose name starts with full stops. If one goes with fullwidth forms for the escapes, `．` (U+FF0E FULLWIDTH FULL STOP) would do the job (this must only be done for the prefix full stops, full stops occuring later in the channel name must not be escaped, to preserve links like https://irclog.whitequark.org/Paws.Nucleus).

whitequark commented:
Something troublesome that I've realized will happen: it will become impossible to type channel names by hand after this change. Hm.

eggrobin commented:
> Something troublesome that I've realized will happen: it will become impossible to type channel names by hand after this change. Hm.

I think that would be a problem with just about any escaping scheme. One could make it variably feasible, e.g. using the character name between commas (comma and space are the only two printable characters disallowed in channel names), but nobody in their right mind should want to type or see `,SOLIDUS,` or `,NUMBER_SIGN,`. Perhaps it could work if a sufficiently palatable comma-alphabetic escape is allowed as a link, but redirects to the more readable fullwidth-escaped address?

I am fairly incompetent here and should get back to principia, I'll try to discuss that further on IRC with @StollD later today (UTC+1). Feel free to drop by if you want to join that discussion—or have a look at [the logs](https://logs.tmsp.io/) :-)

eggrobin commented:
Discussing this further with @StollD (in [#kspunofficial](https://logs.tmsp.io/kspunofficial/2017-03-06)), we are wondering what the criterion is regarding %-encoded atrocity: Unicode trickery will also get %-encoded (more atrociously even), and while a decent browser will display it properly in the address bar, it will be atrocious (again moreso than the original ASCII, e.g. `%EF%BC%8F` instead of `%2F`) in links when copied from there.

whitequark commented:
> Discussing this further with @StollD (in #kspunofficial), we are wondering what the criterion is regarding %-encoded atrocity: Unicode trickery will also get %-encoded (more atrociously even), and while a decent browser will display it properly in the address bar, it will be atrocious (again moreso than the original ASCII, e.g. %EF%BC%8F instead of %2F) in links when copied from there.

If you %-encode the channel name and then put it into the URL, where % is not a valid character, it gets *doubly* percent-encoded, so links will look horrible regardless of which of the two paths we choose. Hm.

I wonder if something like punycode would work here. Let me think about it...

whitequark commented:
Ok, here's the plan.

  * All `#` are replaced with `-h-`.
  * All `.` are replaced with `-d-`.
  * All `/` are replaced with `-s-`.
  * A new configuration option is introduced, that remaps old links before the escaping happens, to preserve them.

eggrobin commented:
You are missing a couple of characters, namely `%`, `\`, and `?`, that also absolutely need to be escaped; maybe `p`, `r`, `q` between `-`s?
Further, the characters  `` `<>|{}^ `` have the interesting property that they appear to work if they are in a link (not sure about `<>` because that would need to be escaped in the html), but if I follow such a link, my browser (chrome 56 here) escapes them itself (and so if copied from the address bar they get percent-escaped). Similarly if I type them into the address bar the browser escapes them.
Thus if we want to avoid percent escapes in links the list of characters to escape is `` %/\?#`<>|{}^ ``.
See below for a proposed set of escapes.

The current implementation (e.g. at https://irclog.whitequark.org) does support percent-escapes, which is why [this commit](https://github.com/whitequark/irclogger/pull/22/commits/d0dfff2e3bfafddf423df250b8b86bd2cd6e7e3b) happened. For instance, this is a valid link to today's logs for `##openfpga`: https://irclog.whitequark.org/%23openfpga/2017-03-07.
The new scheme can (and will likely) continue to do so by ignoring the already de-escaped characters when de-escaping, simply not emitting them in links that are likely to be used.

However, note that a user who is blissfully unaware of all this and types a slightly more reasonable channel name in the address bar, may not end up on the right channel. For instance, if their channel is `#i-like-h-the-constant` they will go to `#i-like#the-constant` (they should have typed `--h--`).
This may not be an issue, however if it is, the character `,` (COMMA) could be used for escapes instead of `-` (HYPHEN-MINUS). Since that may occur in channel names there would be no such collision, and the hypothetical fans of Planck's constant would find their channel.

Apologies for the seemingly unending spam of replies and commits on this one, this is proving trickier than I would have thought. Proposed encoding below.
###### Proposal
```
- → --
# → -h- -- for Hash;
/ → -s- -- for Slash (or Solidus);
% → -p- -- for Percent;
\ → -r- -- for Reverse (slash or solidus);
? → -q- -- for Question mark;
` → -a- -- for grave Accent;
< → -l- -- for Less than;
> → -g- -- for Greater than;
| → -b- -- for vertical Bar;
{ → -o- for Opening curly bracket;
} → -c- for Closing curly bracket;
^ → -x- for circumfleX accent;
```

whitequark commented:
> Thus if we want to avoid percent escapes in links the list of characters to escape is %/\?#`<>|{}^.
> See below for a proposed set of escapes.

Wonderful, I hoped someone would do the legwork for me, because I really didn't want to comb through the spec. Thanks :)

> However, note that a user who is blissfully unaware of all this and types a slightly more reasonable channel name in the address bar, may not end up on the right channel.

This is a good point, `-` is a fairly common character. Let's try a slightly different scheme then: `~` used as an escape character. So for example:

  * `##openfpga` → `~h~openfpga`
  * `#{foobar}` → `~o~foobar~c~`

> This may not be an issue, however if it is, the character , (COMMA) could be used for escapes instead of - (HYPHEN-MINUS). Since that may occur in channel names there would be no such collision, and the hypothetical fans of Planck's constant would find their channel.

Did you mean "may not occur" above? Can `~` occur in channel names? I mildly dislike using comma because it does not obviously connote "escape", whereas tilde is hopefully uncommon enough to do so. But they are both quite unusual for this role.

eggrobin commented:
@whitequark: Uh, right, may not occur. I need another coffee, clearly. I understand not liking `,` for escapes, it's just that the mathematician in me likes injective things (`~` is fine in a channel name, and I agree, there's probably less of a chance of `~h~` being a thing).

Also I nearly typed "I another coffee" above, so I am going to make that coffee right now.

StollD commented:
Implemented the escaping, and added a config option called "legacy". Its a list that stores the values that should get unescaped using the old method.

So adding this to your config should work @whitequark 
`
legacy:
  - ".openfpga"
`

whitequark commented:
1. Indent with two spaces, please.
2. The `legacy` key would be better named `channel_map` or something.
3. That loop looks dreadful, a gsub with a block is a far more elegant solution (try something like `gsub(/~(.)~/) { |m| map_char(m[1]) }` 
4. The two lists of replacements are prone to desync, how about having a `CHANNEL_ESCAPE = { "?" => "q", .. }` then using `CHANNEL_ESCAPE.invert` to go the other way.

StollD commented:
I literally have no experience with ruby, and I only started messing with it because I wanted to correct the issues I / egg discovered. So here are my thoughts on this:

1. sure
2. Well, it isnt a map. It is a list of values where the old (=> legacy) unescaping method gets used, instead of the new system. Maybe `legacy_escaping` or `legacy_channels`?
3. I wont regex this. Sorry that it is ugly, but it is way easier in my opinion
4. Sure
5. Sure

StollD commented:
@whitequark done

whitequark commented on lib/irclogger/viewer.rb:
```diff
@@ -11,7 +11,8 @@
     set :views,         File.expand_path('../../../views', __FILE__)
-    set :public_folder, File.expand_path('../../../public', __FILE__)
+    set :public_folder, File.expand_path('../../../public', __FILE__)    
+    set :protection, :except => :path_traversal 
```
What does this do exactly?

StollD replied:
It was related to the forward slashes in the paths and I forgot it

whitequark commented on lib/irclogger/viewer_helpers.rb:
```diff
@@ -3,13 +3,44 @@
+    end
+    
+    def unescape_url(url)
+      inv = CHANNEL_ESCAPE.invert 
```
This cound be lifted to a constant, to avoid wasting time inverting the hash on every call.

whitequark commented:
Other than the above I'm quite happy with the changes, will probably merge tomorrow.

whitequark commented:
Merged with some minor corrections in master. Note that the config key name was changed.

whitequark commented on lib/irclogger/viewer_helpers.rb:
```diff
@@ -3,13 +3,44 @@
-      "##{channel.gsub(/^\.+/) { |m| '#' * m.length }}"
+      c = unescape_url(channel)
+      if Config.key?('legacy') && Config['legacy'].include?(c) then
+        "##{c.gsub(/^\.+/) { |m| '#' * m.length }}"
```
Uh, this doesn't work, obviously, you have to either add an `else` or `return`.

whitequark commented:
@StollD A few more fixes are in master, and now it's fully usable, and deployed on irclog.whitequark.org.
